pkg/sentry/server/validator/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,262 @@
+		controlPlaneTD: opts.SentryID.TrustDomain(),
+	}, nil
+}
+
+func (k *kubernetes) Start(ctx context.Context) error {
+	if err := k.client.(cache.Cache).Start(ctx); err != nil {
+		return err
+	}
+
+	<-ctx.Done()
+	return nil
+}
+
+func (k *kubernetes) Validate(ctx context.Context, req *sentryv1pb.SignCertificateRequest) (spiffeid.TrustDomain, error) {
+	if !k.ready(ctx) {
+		return spiffeid.TrustDomain{}, errors.New("validator not ready")
+	}
+
+	// The TrustDomain field is ignored by the Kubernetes validator.
+	if _, err := internal.Validate(ctx, req); err != nil {
+		return spiffeid.TrustDomain{}, err
+	}
+
+	prts, err := k.executeTokenReview(ctx, req.GetToken(), LegacyServiceAccountAudience, k.sentryAudience)
+	if err != nil {
+		return spiffeid.TrustDomain{}, err
+	}
+
+	if len(prts) != 4 || prts[0] != "system" {
+		return spiffeid.TrustDomain{}, fmt.Errorf("%s: provided token is not a properly structured service account token", errPrefix)
+	}
+
+	if prts[2] != req.Namespace {
+		return spiffeid.TrustDomain{}, fmt.Errorf("%s: namespace mismatch. received namespace: %s",
+			errPrefix, req.Namespace)
+	}
+
+	// We have already validated to the token against Kubernetes API server, so
+	// we do not need to supply a key.
+	ptoken, err := jwt.ParseInsecure([]byte(req.GetToken()), jwt.WithTypedClaim("kubernetes.io", new(k8sClaims)))
+	if err != nil {
+		return spiffeid.TrustDomain{}, fmt.Errorf("%s: failed to parse Kubernetes token: %s", errPrefix, err)
+	}
+	claimsT, ok := ptoken.Get("kubernetes.io")
+	if !ok {
+		return spiffeid.TrustDomain{}, errMissingPodClaim
+	}
+	claims, ok := claimsT.(*k8sClaims)
+	if !ok || len(claims.Pod.Name) == 0 {
+		return spiffeid.TrustDomain{}, errMissingPodClaim
+	}
+
+	var pod corev1.Pod
+	err = k.client.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: claims.Pod.Name}, &pod)
+	if err != nil {
+		log.Errorf("failed to get pod %s/%s for requested identity: %s", req.Namespace, claims.Pod.Name, err)
+		return spiffeid.TrustDomain{}, fmt.Errorf("%s: failed to get pod of identity", errPrefix)
+	}
+	expID, ok := pod.GetAnnotations()[annotations.KeyAppID]
+	if !ok {
+		expID = pod.GetName()
+	}
+
+	if pod.Spec.ServiceAccountName != prts[3] {
+		log.Errorf("service account on pod %s/%s does not match token", req.Namespace, claims.Pod.Name)
+		return spiffeid.TrustDomain{}, fmt.Errorf("%s: pod service account mismatch", errPrefix)
+	}
+
+	// TODO: @joshvanl: Before v1.11, the injector instructed daprd to request
+	// for the ID containing their namespace and service account (ns:sa). This
+	// is wrong- dapr identities are based on daprd namespace + _app ID_.
+	// Remove this allowance in v1.12.
+	if req.Namespace+":"+pod.Spec.ServiceAccountName == req.Id {
+		req.Id = expID
+	}
+
+	if expID != req.Id {
+		return spiffeid.TrustDomain{}, fmt.Errorf("%s: app-id mismatch. expected: %s, received: %s",
+			errPrefix, expID, req.Id)
+	}
+
+	configName, ok := pod.GetAnnotations()[annotations.KeyConfig]
```

Comment:
When #6101 is merged, there could be more than 1 config referenced. The annotation in this case will be a comma-separated list.

Reply:
What trust domain will the app belong to if multiple configs are assigned? I don't think the current trust domain system is compatible with having multiple config per app

Reply:
Multiple configs works in last-one wins fashion, if more than one config sets the same option.